sdks/java/io/csv/src/main/java/org/apache/beam/sdk/io/csv/CsvIO.java

```diff
@@ -0,0 +1,384 @@
+  public static Write<Row> writeRows() {
+    return new AutoValue_CsvIO_Write.Builder<Row>().build();
+  }
+
+  /** Implementation of {@link FileIO.Sink}. */
+  @AutoValue
+  public abstract static class Sink<T> implements FileIO.Sink<T> {
+
+    public static <T> Builder<T> builder() {
+      return new AutoValue_CsvIO_Sink.Builder<>();
+    }
+
+    /**
+     * Not to be confused with the CSV header, it is content written to the top of every sharded
+     * file prior to the header. In the example below, all the text proceeding the header
+     * 'column1,column2,column3' is the preamble.
+     *
+     * <p>Fake company, Inc. Lab experiment: abcdefg123456 Experiment date: 2022-12-05 Operator:
+     * John Doe
+     *
+     * <p>column1,column2,column3 1,2,3 4,5,6
+     */
+    public Sink<T> withPreamble(String preamble) {
+      return toBuilder().setPreamble(preamble).build();
+    }
+
+    private transient @Nullable PrintWriter writer;
+
+    /**
+     * Opens a {@link WritableByteChannel} for writing CSV files. Writes the {@link #getPreamble()}
+     * if available followed by the {@link #getHeader()}.
+     */
+    @Override
+    public void open(WritableByteChannel channel) throws IOException {
+      writer =
+          new PrintWriter(
+              new BufferedWriter(new OutputStreamWriter(Channels.newOutputStream(channel), UTF_8)));
+      if (getPreamble() != null) {
+        writer.println(getPreamble());
+      }
+      writer.println(getHeader());
+    }
+
+    /** Serializes and writes the {@param element} to a file. */
+    @Override
+    public void write(T element) throws IOException {
+      if (writer == null) {
+        throw new IllegalStateException(
+            String.format("%s writer is null", PrintWriter.class.getName()));
+      }
+      String line = getFormatFunction().apply(element);
+      writer.println(line);
+    }
+
+    @Override
+    public void flush() throws IOException {
+      if (writer == null) {
+        throw new IllegalStateException(
+            String.format("%s writer is null", PrintWriter.class.getName()));
+      }
+      writer.flush();
+    }
+
+    /**
+     * Not to be confused with the CSV header, it is content written to the top of every sharded
+     * file prior to the header. In the example below, all the text proceeding the header
+     * 'column1,column2,column3' is the preamble.
+     *
+     * <p>Fake company, Inc. Lab experiment: abcdefg123456 Experiment date: 2022-12-05 Operator:
+     * John Doe
+     *
+     * <p>column1,column2,column3 1,2,3 4,5,6
+     */
+    abstract @Nullable String getPreamble();
+
+    /**
+     * The column names of the CSV file written at the top line of each shard after the preamble, if
+     * available.
+     */
+    abstract String getHeader();
+
+    /** A {@link SimpleFunction} for converting a {@param T} to a CSV formatted string. */
+    abstract SimpleFunction<T, String> getFormatFunction();
```

Comment:
Use SerializableFunction or ProcessFunction depending on whether you want to allow the user to throw an exception or not.